File: backend/src/main/java/com/backend/petplace/global/s3/S3Uploader.java
package com.backend.petplace.global.s3;

import com.amazonaws.services.s3.AmazonS3Client;
import com.amazonaws.services.s3.model.ObjectMetadata;
import com.backend.petplace.global.exception.BusinessException;
import com.backend.petplace.global.response.ErrorCode;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.UUID;

@Component
public class S3Uploader {

  private final AmazonS3Client amazonS3Client;
  private final String bucket;

  public S3Uploader(AmazonS3Client amazonS3Client, @Value("${cloud.aws.s3.bucket}") String bucket) {
    this.amazonS3Client = amazonS3Client;
    this.bucket = bucket;
  }

  public String upload(MultipartFile multipartFile, String dirName) {
    validateFileName(multipartFile.getOriginalFilename());

    String uniqueFileName = createUniqueFileName(multipartFile.getOriginalFilename(), dirName);
    ObjectMetadata metadata = createObjectMetadata(multipartFile);

    try {
      amazonS3Client.putObject(bucket, uniqueFileName, multipartFile.getInputStream(), metadata);
    } catch (IOException e) {
      throw new BusinessException(ErrorCode.FAIL_TO_UPLOAD_IMAGE, e);
    }
    return amazonS3Client.getUrl(bucket, uniqueFileName).toString();
  }

  private void validateFileName(String originalFilename) {
    if (!StringUtils.hasText(originalFilename)) {
      throw new BusinessException(ErrorCode.INVALID_FILE_NAME);
    }
  }

  private ObjectMetadata createObjectMetadata(MultipartFile multipartFile) {
    ObjectMetadata metadata = new ObjectMetadata();
    metadata.setContentLength(multipartFile.getSize());
    metadata.setContentType(multipartFile.getContentType());
    return metadata;
  }

  private String createUniqueFileName(String originalFilename, String dirName) {
    return dirName + "/" + UUID.randomUUID() + "-" + originalFilename;
  }
}

File: backend/src/main/java/com/backend/petplace/global/s3/S3Uploader.kt
package com.backend.petplace.global.s3

import com.amazonaws.services.s3.AmazonS3
import com.amazonaws.services.s3.model.ObjectMetadata
import com.backend.petplace.global.exception.BusinessException
import com.backend.petplace.global.response.ErrorCode
import org.springframework.beans.factory.annotation.Value
import org.springframework.cglib.core.CollectionUtils.bucket
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils
import org.springframework.web.multipart.MultipartFile
import java.io.IOException
import java.util.*

@Component
class S3Uploader(
    private val amazonS3Client: AmazonS3,

    @Value("\${cloud.aws.s3.bucket}")
    private val bucket: String

) {

    fun upload(multipartFile: MultipartFile, dirName: String): String {
        // 프로퍼티 접근 문법 사용 (.originalFilename)
        val originalFilename = multipartFile.originalFilename
        validateFileName(originalFilename)

        val uniqueFileName = createUniqueFileName(originalFilename!!, dirName)
        val metadata = createObjectMetadata(multipartFile)

        try {
            amazonS3Client.putObject(
                bucket,
                uniqueFileName,
                multipartFile.inputStream,
                metadata
            )
        } catch (e: IOException) {
            throw BusinessException(ErrorCode.FAIL_TO_UPLOAD_IMAGE, e)
        }
        return amazonS3Client.getUrl(bucket, uniqueFileName).toString()
    }

    private fun validateFileName(originalFilename: String?) {
        if (originalFilename.isNullOrBlank()) {
            throw BusinessException(ErrorCode.INVALID_FILE_NAME)
        }
    }

    private fun createObjectMetadata(multipartFile: MultipartFile): ObjectMetadata {

        return ObjectMetadata().apply {
            contentLength = multipartFile.size       // .getSize() -> .size
            contentType = multipartFile.contentType  // .getContentType() -> .contentType
        }
    }

    private fun createUniqueFileName(originalFilename: String, dirName: String): String {
        // 문자열 템플릿 사용 ("$변수")
        return "$dirName/${UUID.randomUUID()}-$originalFilename"
    }
}